tests/e2e/set-ffi-env-e2e

```diff
@@ -154,140 +167,313 @@ while [ $# -gt 0 ]; do
     esac
 done
 
-setup_qm_services() {
+# Utility function to create configuration files
+create_config_file() {
+    local file_path="$1"
+    local content="$2"
+    local description="$3"
+
+    if [[ -n "$description" ]]; then
+        info_message "Creating $description..."
+    fi
+
+    # Create directory if it doesn't exist
+    local dir_path
+    dir_path=$(dirname "$file_path")
+    mkdir -p "$dir_path"
+
+    # Create the file with content
+    printf "%s\n" "$content" > "$file_path"
+    if [[ -n "$description" ]]; then
+        info_message "$description created successfully at $file_path"
+    fi
+}
+
+# Utility function to append content to existing files
+append_to_config_file() {
+    local file_path="$1"
+    local content="$2"
+    local description="$3"
+
+    if [[ ! -f "$file_path" ]]; then
+        info_message "Warning: Configuration file not found: $file_path"
+        return 1
+    fi
+
+    # Check if content already exists to avoid duplicates
+    if grep -Fq "$content" "$file_path"; then
+        info_message "$description already exists in $file_path"
+        return 0
+    fi
+
+    if [[ -n "$description" ]]; then
+        info_message "Appending $description to $file_path..."
+    fi
+
+    echo "$content" >> "$file_path"
+
+    if [[ -n "$description" ]]; then
+        info_message "$description appended successfully"
+    fi
+}
+
+# Utility function to display section headers
+section_header() {
+    local title="$1"
+    local separator="${2:-==============================}"
+
+    info_message "$title"
+    info_message "$separator"
+}
+
+# Utility functions for systemctl operations
+enable_services() {
+    local services=("$@")
+    for service in "${services[@]}"; do
+        info_message "Enabling service: $service"
+        systemctl enable "$service"
+    done
+}
+
+start_services() {
+    local services=("$@")
+    for service in "${services[@]}"; do
+        info_message "Starting service: $service"
+        systemctl start "$service"
+    done
+}
+
+restart_services() {
+    local services=("$@")
+    for service in "${services[@]}"; do
+        info_message "Restarting service: $service"
+        systemctl restart "$service"
+    done
+}
```

Comment:
These are very general functions and could be used in other locations as well. So probably good to move to something like ` tests/e2e/lib/systemd` (at least the file path/name suggest that it could fit there).